File: 5-Strings/LSD.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class LSD {
    private static int R = 256;
    /**
     * Least-Significant-Digit First sort.
     *
     * @param a the array of string to be sorted
     * @param W the number of characters per string
     */
    public static void sort(String[] a, int W) {
        int N = a.length;
        String[] aux = new String[N];

        for (int d = W-1; d >= 0; d--) {
            int[] count = new int[R+1];
            // frequency counts
            for (int i = 0; i < N; i++) {
                count[a[i].charAt(d)+1]++;
            }

            // cumulative counts
            for (int i = 0; i < R; i++) {
                count[i+1] += count[i];
            }

            // move to aux
            for (int i = 0; i < N; i++) {
                aux[count[a[i].charAt(d)]++] = a[i];
            }

            // copy back
            for (int i = 0; i < N; i++) {
                a[i] = aux[i];
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());  // length of the array
        int W = Integer.parseInt(br.readLine());  // width of each string
        String[] a = new String[N];  // array of strings
        for (int i = 0; i < N; i++) {
            a[i] = br.readLine();
        }
        LSD.sort(a, W);
        for (int i = 0; i < N; i++) {
            System.out.print(i);
            System.out.print(" " + a[i] + "\n");
        }
    }
}

File: 5-Strings/LSD.cpp
#include <iostream>
#include <string>
#include <vector>

using std::cout;
using std::cin;
using std::cerr;
using std::endl;
using std::vector;
using std::string;

static const int R = 256;

static void LSD_sort(vector<string>& a, int width)
{
    vector<string> aux(a.size());
    for (size_t i = 0; i < a.size(); i++) {
        aux[i] = a[i];
    }

    for (int d = width-1; d >= 0; --d) {
        int count[R+1] = {0};
        // 1-counts
        for (size_t i = 0; i < a.size(); i++) {
            ++count[static_cast<int>(a[i].at(d))+1];
        }

        // 2-cumulative counts
        for (int i = 0; i < R; i++) {
            count[i+1] += count[i];
        }

        // 3-move to aux
        for (size_t i = 0; i < a.size(); i++) {
            aux[count[static_cast<int>(a[i].at(d))]++] = a[i];
        }

        // 4-copy back
        for (size_t i = 0; i < a.size(); i++) {
            a[i] = aux[i];
        }
    }
}


int main(int argc, char* argv[])
{
    vector<string> a;
    string s;
    cin >> s;
    a.push_back(s);
    size_t width = s.size();
    while (cin >> s) {
        if (s.size() != width) {
            cerr << "Unequal size of strings" << endl;
            exit(1);
        }
        a.push_back(s);
    }

    LSD_sort(a, width);

    for (size_t i = 0; i < a.size(); i++) {
        cout << i << " " << a[i] << endl;
    }

    return 0;
}
